README.md
# generator-laravel-5 [![NPM version][npm-image]][npm-url] [![Build Status][travis-image]][travis-url] [![Dependency Status][daviddm-image]][daviddm-url]
> Installs Laravel-5 with a bunch of useful packages and enables BrowserSync

## Installation

First, install [Yeoman](http://yeoman.io) and generator-laravel-5 using [npm](https://www.npmjs.com/) (we assume you have pre-installed [node.js](https://nodejs.org/)).

```bash
npm install -g yo
npm install -g generator-laravel-5
```

Then generate your new project:

```bash
yo laravel-5
```

## Proxy
You can choose from where your application is served:
* php artisan serve (localhost:8000)
* localhost (Use this, if you serve your application via a webserver (local or remote))
The proxy configuration is required for BrowserSync and can be changed in `webpack.mix.js`

## Integrated packages
The following packages are integrated in the new Laravel project by default:
* doctrine/dbal
* barryvdh/laravel-ide-helper
* barryvdh/laravel-debugbar
* barryvdh/laravel-cors
* phpmetrics/phpmetrics
* beyondcode/laravel-self-diagnosis
* symplify/easy-coding-standard

## phpmetrics analyzes
In order to perform a code analyzes with phpmetrics you can run:
```
composer run-script analyze
```
This will create a folder named `phpmetrics` and you can get interesting insights by opening
`phpmetrics/index.html`

## Code style check and fixing
In order to have a clean and shiny code style, your generated Laravel 5 application comes with 
[EasyCodingStandard](https://github.com/Symplify/EasyCodingStandard) support.  
You can check your code with the following command: `composer run-script ecsCheck`
This checks the whole project and applies some basic check configurations.
You can tweak the check configuration in the `easy-coding-standard.yml` file.

In order to fix your code style, you can run `composer run-script ecsFix`.

You can also run easyCodingStandard and pass other configuration by calling the tool directly:
```
vendor/bin/ecs check app --config vendor/symplify/easy-coding-standard/config/clean-code.yml
vendor/bin/ecs check app --config vendor/symplify/easy-coding-standard/config/clean-code.yml --fix
```

For more information, please consult the official 
[documentation](https://github.com/Symplify/EasyCodingStandard/blob/master/README.md).

## phpunit-watcher
Another useful php package that is integrated out of the box is spatie/phpunit-watcher. With this tool, 
you can watch changes in *.php and test files and all your automated tests will be triggered to run.
You can start phpunit-watcher with:
```bash
vendor/bin/phpunit-watcher watch
```
For more information have a look at the official documentation:
https://github.com/spatie/phpunit-watcher

## Frontend development with browserSync
If you serve your application with `php artisan serve` (http://localhost:8000) you have to start
this first before you can start watching your files with browserSync.

In order to have webpack watch your file changes and reload your browser, run `npm run watch`.
You have to open your browser an navigate to `http:localhost:3000`

## Git versioning
If you want to, you can initialize a local git repository to version your code.
This is turned on by default. After scaffolding your new laravel application there is an initial
commit in your local repository.

## License

Apache-2.0 © [Sebastian Baum](http://www.sebbaum.de)

[npm-image]: https://badge.fury.io/js/generator-laravel-5.svg
[npm-url]: https://npmjs.org/package/generator-laravel-5
[travis-image]: https://travis-ci.org/sebbaum/generator-laravel-5.svg?branch=master
[travis-url]: https://travis-ci.org/sebbaum/generator-laravel-5
[daviddm-image]: https://david-dm.org/sebbaum/generator-laravel-5.svg?theme=shields.io
[daviddm-url]: https://david-dm.org/sebbaum/generator-laravel-5


CHANGELOG.md
# CHANGELOG

## 1.7.0
* Integrated laravel/telescope for Laravel 5.7+ applications

## 1.6.0
* Integrated spatie/phpunit-watcher

## 1.5.0
* Generating ide-helper files.
* Creating a `database.sqlite` file in the database folder.
* Integrated [EasyCodingStandard](https://github.com/Symplify/EasyCodingStandard) to check and fix the code style.

## 1.4.0
* Optional initialization of a local git repository.
* Optimized documentation and code comments.
* Optional integration of Laravel's auth layer.

## 1.3.0
* Fixed missing NPM dependencies in the context of frontend presets
* Fixed the php artisan serve port in webpack.mix.js
* Added more automated tests

## 1.2.0
* Supporting nodejs 10

## 1.1.0
* Added support for Laravel 5.7.*
* Linking storage to public folder

## 0.5.0
* Improved support for local or remote development.

## 0.4.0
* Added phpmetrics to the generator
* Intrduced the CHANGELOG file
* Added more automated tests

## 0.3.0
* Added automated tests

## 0.2.0
* Removed the fix application path. Now there is more flexibility.

## 0.1.0
* Basic implemenation of the generator

__tests__/app.js
'use strict';
const path = require('path');
const assert = require('yeoman-assert');
const helpers = require('yeoman-test');

describe('generator-laravel-5:app', () => {
  beforeEach(done => {
    helpers
      .run(path.join(__dirname, '../generators/app'))
      .withPrompts({
        appname: 'testapp',
        version: '',
        proxy: 'localhost',
        schema: 'http',
        preset: 'none'
      })
      .on('end', done);
  }, 1200000);

  it('creates files', done => {
    assert.file(['webpack.mix.js', 'package.json', '.gitignore']);
    done();
  });

  it('composer.json contains laravel packages and script', done => {
    assert.fileContent('composer.json', 'doctrine/dbal');
    assert.fileContent('composer.json', 'barryvdh/laravel-ide-helper');
    assert.fileContent('composer.json', 'barryvdh/laravel-debugbar');
    assert.fileContent('composer.json', 'barryvdh/laravel-cors');
    assert.fileContent('composer.json', 'phpmetrics/phpmetrics');
    assert.fileContent('composer.json', 'beyondcode/laravel-self-diagnosis');
    assert.jsonFileContent('composer.json', {
      scripts: {
        analyze: ['phpmetrics --report-html=phpmetrics ./app']
      }
    });

    done();
  });
});


__tests__/app-5-5.js
'use strict';
const path = require('path');
const assert = require('yeoman-assert');
const helpers = require('yeoman-test');

describe('generator-laravel-5:app', () => {
  beforeEach(done => {
    helpers
      .run(path.join(__dirname, '../generators/app'))
      .withPrompts({
        appname: 'testapp',
        version: '5.5.*',
        proxy: 'localhost',
        schema: 'http',
        preset: 'none',
        enableAuth: false,
        localGit: false
      })
      .on('end', done);
  }, 1200000);

  it('creates files', done => {
    assert.file([
      'webpack.mix.js',
      'package.json',
      '.gitignore',
      // '.phpstorm.meta.php',
      // '_ide_helper.php',
      'database/database.sqlite',
      'easy-coding-standard.yml'
    ]);
    done();
  });

  it('composer.json contains laravel packages and script', done => {
    assert.fileContent('composer.json', 'doctrine/dbal');
    assert.fileContent('composer.json', 'barryvdh/laravel-ide-helper');
    assert.fileContent('composer.json', 'barryvdh/laravel-debugbar');
    assert.fileContent('composer.json', 'barryvdh/laravel-cors');
    assert.fileContent('composer.json', 'phpmetrics/phpmetrics');
    assert.fileContent('composer.json', 'beyondcode/laravel-self-diagnosis');
    assert.fileContent('composer.json', 'symplify/easy-coding-standard');
    assert.fileContent('composer.json', 'spatie/phpunit-watcher');
    assert.jsonFileContent('composer.json', {
      scripts: {
        analyze: ['phpmetrics --report-html=phpmetrics ./app'],
        ecsCheck: ['vendor/bin/ecs check .'],
        ecsFix: ['vendor/bin/ecs check . --fix']
      }
    });

    done();
  });

  it('webpack.mix.js is configured correctly', done => {
    assert.fileContent(
      'webpack.mix.js',
      "mix.js('resources/assets/js/app.js', 'public/js');"
    );
    assert.fileContent(
      'webpack.mix.js',
      "mix.sass('resources/assets/sass/app.scss', 'public/css');"
    );
    assert.fileContent(
      'webpack.mix.js',
      'mix.browserSync({\n' +
        "  proxy: 'http://localhost',\n" +
        "  host: 'localhost',\n" +
        '  open: false,\n' +
        '  watchOptions: {\n' +
        '    usePolling: true\n' +
        '  }\n' +
        '});'
    );

    done();
  });
});


__tests__/app-5-6.js
'use strict';
const path = require('path');
const assert = require('yeoman-assert');
const helpers = require('yeoman-test');

describe('generator-laravel-5:app', () => {
  beforeEach(done => {
    helpers
      .run(path.join(__dirname, '../generators/app'))
      .withPrompts({
        appname: 'testapp',
        version: '5.6.*',
        proxy: 'localhost',
        schema: 'http',
        preset: 'none',
        enableAuth: false,
        localGit: false
      })
      .on('end', done);
  }, 1200000);

  it('creates files', done => {
    assert.file([
      'webpack.mix.js',
      'package.json',
      '.gitignore',
      // '.phpstorm.meta.php',
      // '_ide_helper.php',
      'database/database.sqlite',
      'easy-coding-standard.yml'
    ]);
    done();
  });

  it('composer.json contains laravel packages and script', done => {
    assert.fileContent('composer.json', 'doctrine/dbal');
    assert.fileContent('composer.json', 'barryvdh/laravel-ide-helper');
    assert.fileContent('composer.json', 'barryvdh/laravel-debugbar');
    assert.fileContent('composer.json', 'barryvdh/laravel-cors');
    assert.fileContent('composer.json', 'phpmetrics/phpmetrics');
    assert.fileContent('composer.json', 'beyondcode/laravel-self-diagnosis');
    assert.fileContent('composer.json', 'symplify/easy-coding-standard');
    assert.fileContent('composer.json', 'spatie/phpunit-watcher');
    assert.jsonFileContent('composer.json', {
      scripts: {
        analyze: ['phpmetrics --report-html=phpmetrics ./app'],
        ecsCheck: ['vendor/bin/ecs check .'],
        ecsFix: ['vendor/bin/ecs check . --fix']
      }
    });

    done();
  });

  it('webpack.mix.js is configured correctly', done => {
    assert.fileContent(
      'webpack.mix.js',
      "mix.js('resources/assets/js/app.js', 'public/js');"
    );
    assert.fileContent(
      'webpack.mix.js',
      "mix.sass('resources/assets/sass/app.scss', 'public/css');"
    );
    assert.fileContent(
      'webpack.mix.js',
      'mix.browserSync({\n' +
        "  proxy: 'http://localhost',\n" +
        "  host: 'localhost',\n" +
        '  open: false,\n' +
        '  watchOptions: {\n' +
        '    usePolling: true\n' +
        '  }\n' +
        '});'
    );

    done();
  });
});


generators/app/templates/webpack.mix.js
let mix = require('laravel-mix');

/*
 |--------------------------------------------------------------------------
 | Mix Asset Management
 |--------------------------------------------------------------------------
 |
 | Mix provides a clean, fluent API for defining some Webpack build steps
 | for your Laravel application. By default, we are compiling the Sass
 | file for the application as well as bundling up all the JS files.
 |
 */
<% if (version ==='5.7.*') { %>
mix.js('resources/js/app.js', 'public/js')
  .sass('resources/sass/app.scss', 'public/css');
<% } else { %>
  mix.js('resources/assets/js/app.js', 'public/js')
    .sass('resources/assets/sass/app.scss', 'public/css');
<% } %>


/*
|-----------------------------------------------------------------------
| BrowserSync
|-----------------------------------------------------------------------
|
| BrowserSync refreshes the Browser if file changes (js, sass, php) are | detected.
| Proxy specifies the location where the app is located.
*/
mix.browserSync({
  proxy: '<%= proxy %>',
  host: 'dev.host',
  open: false,
  watchOptions: {
    usePolling: true
  }
});

generators/app/index.js
'use strict';
const Generator = require('yeoman-generator');
const chalk = require('chalk');
const fs = require('fs');

module.exports = class extends Generator {
  /**
   * Ask the developer some questions that will define the build result.
   * @returns {Promise|*|PromiseLike<T | never>|Promise<T | never>}
   */
  prompting() {
    this.log("Let's create a new " + chalk.blue('Laravel 5') + ' application');

    const questions = [
      {
        type: 'input',
        name: 'appname',
        message: "What's the name of your application?",
        default: 'application'
      },
      {
        type: 'list',
        name: 'version',
        message: 'Which Laravel version do you want to use?',
        choices: ['5.7.*', '5.6.*', '5.5.*'],
        default: 0
      },
      {
        type: 'list',
        name: 'proxy',
        message: 'From where do you serve your application during development?',
        choices: ['php artisan serve (localhost:8000)', 'localhost'],
        default: 0
      },
      {
        when: answers => {
          return answers.proxy === 'localhost';
        },
        type: 'list',
        name: 'schema',
        message: 'Which schema do you want to use?',
        choices: ['http', 'https'],
        default: 'https'
      },
      {
        type: 'list',
        name: 'preset',
        message: 'Which laravel frontend preset do you want to use?',
        choices: ['none', 'bootstrap', 'vue', 'react'],
        default: 'none'
      },
      {
        type: 'confirm',
        name: 'enableAuth',
        message: "Do you want to create Laravel's auth layer?",
        default: false
      },
      {
        type: 'confirm',
        name: 'localGit',
        message: 'Do you want to init a local git repository?',
        default: true
      }
    ];

    return this.prompt(questions).then(answers => {
      // To access answers later use this.answers.someAnswer;
      this.answers = answers;
    });
  }

  /**
   * Given the appname and version let's create a new Laravel 5.* application
   */
  installLaravel() {
    this.spawnCommandSync('composer', [
      'create-project',
      '--prefer-dist',
      'laravel/laravel',
      this.answers.appname,
      this.answers.version
    ]);
  }

  /**
   * Define the destination root folder. This will be used in the following build steps.
   */
  setApplicationFolder() {
    this.destinationRoot(this.destinationPath(this.answers.appname));
  }

  /**
   * Install some third party laravel packages that might be useful for your application.
   */
  installLaravelPackages() {
    this.spawnCommandSync('composer', [
      'require',
      '--dev',
      'doctrine/dbal',
      'barryvdh/laravel-ide-helper',
      'barryvdh/laravel-debugbar',
      'barryvdh/laravel-cors',
      'phpmetrics/phpmetrics',
      'beyondcode/laravel-self-diagnosis',
      'symplify/easy-coding-standard',
      'spatie/phpunit-watcher'
    ]);
  }

  /**
   * Add additional composer scripts to composer.json. Composer.json is deleted and then written
   * with the new scripts definition.
   */
  composerScripts() {
    let data = fs.readFileSync('composer.json', 'utf8');
    let composer = JSON.parse(data);
    composer.scripts.analyze = ['phpmetrics --report-html=phpmetrics ./app'];
    composer.scripts.ecsCheck = ['vendor/bin/ecs check .'];
    composer.scripts.ecsFix = ['vendor/bin/ecs check . --fix'];

    fs.unlinkSync('composer.json');

    this.fs.write(
      this.destinationPath('composer.json'),
      JSON.stringify(composer, null, 2)
    );
  }

  /**
   * Run `php artisan preset` with the given preset. This will prepare the frontend files and structure.
   */
  preset() {
    this.spawnCommandSync('php', ['artisan', 'preset', this.answers.preset]);
  }

  /**
   * Remove certain files that will be created by the generator in the following build steps.
   */
  removeFiles() {
    fs.unlinkSync('webpack.mix.js');
    fs.unlinkSync('package.json');
    fs.unlinkSync('.gitignore');
  }

  /**
   * Create new files from templates.
   * - webpack.mix.ejs => webpack.mix.js
   * - package.ejs => package.js
   * - .gitignore => .gitignore
   */
  templates() {
    this.fs.write(this.destinationPath('database/database.sqlite'), '');
    let schema = this.answers.schema || 'http';
    let proxyHost =
      this.answers.proxy === 'localhost' ? this.answers.proxy : 'localhost:8000';
    let proxy = schema + '://' + proxyHost;

    let resourcesPath =
      this.answers.version === '5.7.*' ? 'resources' : 'resources/assets';
    this.fs.copyTpl(
      this.templatePath('webpack.mix.ejs'),
      this.destinationPath('webpack.mix.js'),
      {
        version: this.answers.version,
        preset: this.answers.preset,
        proxy: proxy,
        resourcesPath: resourcesPath
      }
    );

    this.fs.copyTpl(
      this.templatePath('package.ejs'),
      this.destinationPath('package.json'),
      {
        preset: this.answers.preset
      }
    );

    this.fs.copy(this.templatePath('gitignore'), this.destinationPath('.gitignore'));
    this.fs.copy(
      this.templatePath('easy-coding-standard.yml'),
      this.destinationPath('easy-coding-standard.yml')
    );
  }

  /**
   * Perform the frontend installation with NPM.
   * Yarn and bower are disabled.
   */
  install() {
    this.installDependencies({
      bower: false,
      yarn: false,
      npm: true
    });
  }

  /**
   * Finish the build by running php artisan commands and npm scripts.
   * After this step your new application is ready.
   */
  end() {
    if (this.answers.preset !== 'none') {
      this.spawnCommandSync('npm', ['run', 'dev']);
    }
    if (this.answers.enableAuth) {
      this.spawnCommandSync('php', ['artisan', 'make:auth']);
    }

    this.spawnCommandSync('php', ['artisan', 'storage:link']);
    this.spawnCommandSync('php', ['artisan', 'ide-helper:generate']);
    this.spawnCommandSync('php', ['artisan', 'ide-helper:meta']);
    if (this.answers.localGit) {
      this.spawnCommandSync('git', ['init']);
      this.spawnCommandSync('git', ['add', '.']);
      this.spawnCommandSync('git', [
        'commit',
        '-m',
        'Initial commit by yeoman laravel-5 generator'
      ]);
    }
    this.spawnCommandSync('php', ['artisan', 'self-diagnosis']);
  }
};
